atoma-service/src/middleware.rs

```diff
@@ -334,8 +335,24 @@ pub async fn verify_stack_permissions(
             StatusCode::UNAUTHORIZED
         })?;
     if available_stack.is_none() {
-        error!("No available stack with enough compute units");
-        return Err(StatusCode::UNAUTHORIZED);
+        let tx_digest_str = req_parts
+            .headers
+            .get("X-Tx-Digest")
+            .ok_or_else(|| {
+                error!("Tx digest header not found");
+                StatusCode::BAD_REQUEST
+            })?
+            .to_str()
+            .map_err(|_| {
+                error!("Tx digest cannot be converted to a string");
+                StatusCode::BAD_REQUEST
+            })?;
+        let tx_digest = TransactionDigest::from_str(tx_digest_str).unwrap();
+        let compute_units = utils::request_blockchain_for_stack(&state, tx_digest).await?;
+        if compute_units > total_num_compute_units as u64 {
```

Comment:
you have to verify that the `stack_small_id` in the request matches the `stack_small_id` in the tx digest.

Reply:
Great catch !